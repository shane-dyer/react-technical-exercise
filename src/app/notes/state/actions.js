import { loadNotesRequest, addNoteRequest } from './service';

const ns = 'notes/';
export const LOAD_NOTES = `${ns}LOAD_NOTES`;
export const LOAD_NOTES_SUCCESS = `${ns}LOAD_NOTES_SUCCESS`;

export const ADD_NOTE = `${ns}ADD_NOTE`;
export const ADD_NOTE_SUCCESS = `${ns}ADD_NOTE_SUCCESS`;

export function loadNotes() {
  return {
    type: LOAD_NOTES,
    noteOperation: {
      request: loadNotesRequest(),
      success: loadNotesSuccess,
    },
  };
}

function loadNotesSuccess(notes) {
  return {
    type: LOAD_NOTES_SUCCESS,
    notes,
  };
}

export function addNote(text) {
  return {
    type: ADD_NOTE,
    noteOperation: {
      request: addNoteRequest(text),
      success: addNoteSuccess,
    },
  };
}

function addNoteSuccess(note) {
  return {
    type: ADD_NOTE_SUCCESS,
    note,
  };
}
