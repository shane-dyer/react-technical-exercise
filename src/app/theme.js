const global = {
  space: [0, 4, 8, 16, 32],
  fontSizes: {
    alpha: 48,
    beta: 32,
    gamma: 24,
    epsilon: 16,
  },
};

const lightTheme = {
  ...global,
  pageBackground: 'papayawhip',
  pageTitle: 'blueViolet',
  pageText: 'slategrey',
  noteBackground: '#f8f8f8',
  noteHoverBackground: '#fbe2e2',
  noteBorderStyle: '1px solid rgba(0,0,0,0.1)',
  label: 'slateGrey',
  textFieldBackground: '#f8f8f8',
  textFieldFocusBackground: 'lightsteelblue',
  textFieldBorderStyle: '0',
  buttonText: 'white',
  buttonBackground: 'blueViolet',
  buttonDeleteBackground: 'red',
  buttonThemeBackground: 'rgba(0,0,0, 0.06)',
  addNoteBackground: 'rgba(0,0,0, 0.05)',
};

const darkTheme = {
  ...global,
  pageBackground: '#222',
  pageTitle: 'blueViolet',
  pageText: '#ddd',
  noteBackground: '#333',
  noteHoverBackground: 'rgba(255,255,255, 0.08)',
  noteBorderStyle: '1px solid rgba(255,255,255,0.1)',
  label: '#ddd',
  textFieldBackground: '#f8f8f8',
  textFieldFocusBackground: 'lightsteelblue',
  textFieldBorderStyle: '0',
  buttonText: 'white',
  buttonBackground: 'blueViolet',
  buttonDeleteBackground: 'red',
  buttonThemeBackground: 'rgba(255,255,255, 0.15)',
  addNoteBackground: 'rgba(255,255,255, 0.08)',
};

export { lightTheme, darkTheme };
