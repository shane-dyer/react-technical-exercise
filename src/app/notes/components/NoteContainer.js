import { connect } from 'react-redux';
import { updateNote, deleteNote } from '../state';
import Note from './Note';

function mapStateToProps(state, ownProps) {
  return {};
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    updateNote: (note) => {
      dispatch(updateNote(note));
    },
    deleteNote: (id) => {
      dispatch(deleteNote(id));
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Note);
