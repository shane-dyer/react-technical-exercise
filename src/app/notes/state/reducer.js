import { LOAD_NOTES, LOAD_NOTES_SUCCESS, ADD_NOTE, ADD_NOTE_SUCCESS } from './actions';

const initialState = {
  notes: {},
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_NOTES:
      return state;
    case LOAD_NOTES_SUCCESS:
      return {
        ...state,
        notes: action.notes,
      };
    case ADD_NOTE:
      return state;
    case ADD_NOTE_SUCCESS:
      return {
        ...state,
        notes: {
          ...state.notes,
          [action.note.id]: action.note,
        },
      };
    default:
      return state;
  }
}
