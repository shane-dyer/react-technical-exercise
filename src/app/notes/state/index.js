import { loadNotes, addNote, deleteNote } from './actions';
import reducer from './reducer';
import epic from './epic';

export { reducer, epic, loadNotes, addNote, deleteNote };
