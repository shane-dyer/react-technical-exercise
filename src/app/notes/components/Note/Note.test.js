import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import NoteContainer from './NoteContainer';
import { deleteNote } from '../../state';
import store from '../../../store';
import { lightTheme } from '../../../theme';

const note = {
  id: 1,
  text: 'Test note 1',
};

describe('Note component', () => {
  let component;

  beforeEach(() => {
    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <ThemeProvider theme={lightTheme}>
          <NoteContainer note={note} />
        </ThemeProvider>
      </Provider>
    );
  });

  it('should render with given state from redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render with passed props', () => {
    expect(component.root.findByType(NoteContainer).props.note.text).toBe(note.text);
  });

  it('should dispatch an action on delete button click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(deleteNote(note.id));
  });
});
