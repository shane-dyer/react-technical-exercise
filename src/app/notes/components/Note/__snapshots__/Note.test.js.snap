// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note component should render with given state from redux store 1`] = `
<div
  className="sc-bdVaJa bXxAPE"
>
  <p
    onClick={[Function]}
  >
    Test note 1
  </p>
  <button
    aria-label="Delete note"
    className="delete-button"
    onClick={[Function]}
    title="Delete note"
    type="button"
  >
    Delete
  </button>
</div>
`;
